### PR TITLE
Replace GitHub cache library with our own implementation

### PR DESCRIPTION
**Pull request by aomarks**
Replaces [`@actions/cache`](https://github.com/actions/toolkit/tree/main/packages/cache) with our own implementation for interacting with the GitHub Actions cache service.

### API mismatch with @actions/cache

`@actions/cache` takes a key and a set of glob patterns, and the literal glob patterns themselves are automatically used to generate a cache "version". The key + version then serve as a compound key for the cache entry.

This didn't work well for us because the glob implementation it uses doesn't support exclusions the way we need, so it is incompatible with our glob behavior, and may also be different in other subtle ways. We really just want to provide a list of files/directories for the tarball, and set the key independently.

But if we list all the files directly, then we can never get a cache hit, because the full list of files needs to be part of the cache key, but knowing the list of files requires running the script, which is what we're trying to avoid by restoring from cache!

### The workaround we had before

The workaround we had before was to reach into the `internal/` directory of `@actions/cache` and call some of the lower level functions directly, which let us control exactly what the cache key + version was, while still passing an explicit list of files to `tar`.

This worked ok, but had a high risk of breakage, because it uses non-public APIs. https://github.com/google/wireit/pull/227 is an example of us breaking because of that. It also required us to write some of our own typings, and to turn off `skipLibCheck` in our `tsconfig.json`. It was still quite a lot of code, too.

In addition, we still had to deal with an edge case relating to empty directories. By default, `tar` includes all recursive contents of a directory -- so if we wanted to cache an empty directory that wasn't actually empty on disk, we had to resort to a weird "empty directories manifest" hack. But now we can just use `tar --no-recursion` flag which avoids that whole issue.

### This PR

This PR completely replaces `@actions/cache` with a custom implementation that uses `https.request` and `execFile('tar', ...)`.

As a big bonus, **our install size decreased from 25MB to 2.4MB** and our **transitive dependency count decreased from 93 to 29**.

### Testing

- In https://github.com/google/wireit/pull/232 I improved our fake in a number of ways to make it higher fidelity, such as validating headers, and handling multiple upload chunks.

- I read through the implementation details in https://github.com/actions/toolkit/tree/main/packages/cache to make sure we are compatible, picking the same chunk sizes, etc.

- https://github.com/google/wireit/actions/runs/2325580687 is a run that tests this PR against the live servers by installing from an `npm pack`ed version. https://github.com/google/wireit/actions/runs/2325616423 is a follow up re-run, to show that everything got restored from cache.

Fixes https://github.com/google/wireit/issues/107

### CODE REVIEW

File: src/caching/github-actions-cache.ts [hunk 1]
```diff
@@ -184,123 +152,182 @@ export class GitHubActionsCache implements Cache {
...
+  }
 
-    const compressionMethod = await GitHubActionsCache.#compressionMethod;
-    const version = this.#computeVersion(stateStr, compressionMethod);
-
-    let emptyDirsManifestPath: string | undefined;
-    if (emptyDirs.size > 0) {
-      const emptyDirsManifest = JSON.stringify([...emptyDirs]);
-      emptyDirsManifestPath = this.#emptyDirectoriesManifestPath(
+  async #reserveUploadAndCommitTarball(
```
Comment (rictic) [inline src/caching/github-actions-cache.ts:171]: Document what the boolean return means

File: src/caching/github-actions-cache.ts [hunk 2]
```diff
@@ -184,123 +152,182 @@ export class GitHubActionsCache implements Cache {
...
+      return false;
+    }
+    if (!(await this.#commit(script, id, tarballBytes))) {
+      return false;
+    }
+    return true;
+  }
 
-    const tarballPath = await this.#makeTarball([...files], compressionMethod);
+  async #upload(
```
Comment (rictic) [inline src/caching/github-actions-cache.ts:217]: Ditto

File: src/caching/github-actions-cache.ts [hunk 3]
```diff
@@ -184,123 +152,182 @@ export class GitHubActionsCache implements Cache {
...
-      await tarballDeleted;
+      await tarballHandle.close();
+    }
+  }
+
+  async #commit(
+    script: ScriptReference,
+    id: number,
+    tarballBytes: number
+  ): Promise<boolean> {
```
Comment (rictic) [inline src/caching/github-actions-cache.ts:282]: ditto

**Review by rictic — APPROVED** [review]
Excellent comments, great work validating this

File: src/caching/github-actions-cache.ts (hunk 1 above)
Comment (aomarks) [reply src/caching/github-actions-cache.ts:176]: Done

File: src/caching/github-actions-cache.ts (hunk 3 above)
Comment (aomarks) [reply src/caching/github-actions-cache.ts:295]: Done

File: src/caching/github-actions-cache.ts (hunk 2 above)
Comment (aomarks) [reply src/caching/github-actions-cache.ts:226]: Done

**Pull request merged**